### PR TITLE
ORTModule log clean up

### PR DESCRIPTION
**Pull request by pengwa**
### ORTModule log clean up

ORTModule log level - WARNING(Default) is for end users; INFO and VERBOSE is for internal ORT training developers.

Few issues: 
1. ONNX export will output lots of WARNING error message like "The shape inference of com.microsoft::SoftmaxCrossEntropyLossInternal/ATen/PythonOp type is missing", which is useless for us or end users.
image

3. ORT also print some information like ""CleanUnusedInitializersAndNodeArgs] Removing initializer","ReverseBFSWithStopGradient] Skip building gradient for", which is also useless for us or end users most of the time. 
image

5. Different ranks output logs and making ORT developers or end users feels there are too many logs but usually not useful until we need investigate. 

Few improvements for the issues:
1. For ONNX export logs, there are two kinds of logs: a. export verbose log; b. other logs printed by torch C++ backend. So this PR make following change:
        # VERBOSE -> FULL export verbose log + FULL torch other logs from stdout and stderr (C++ backend)
        # INFO -> FULL export verbose log + FILTERED torch other logs from stdout and stderr (C++ backend)
        # WARNING/ERROR -> [Rank 0] NO export verbose log + FILTERED torch other logs from stdout and stderr (C++ backend)

   e.g. for verbose level, print all logs as usually; for info level, print verbose export log, and filtered logs from torch C++ backend (removing messages like this "The shape inference of com.microsoft::SoftmaxCrossEntropyLossInternal/ATen/PythonOp type is missing") . For higher level, only log the info on rank 0. 

2. For ORT gradient graph build and session creation, also suppress the message and filtered out the message when log level >=INFO. 

3. log level > INFO, then only logs on rank 0 is logged, to have a cleaner user experience

This is the log for a BLOOM model training after the change: there are limited of warnings. 

image

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_logger.py [hunk 1]
```diff
@@ -23,31 +26,66 @@ class LogLevel(IntEnum):
...
+    def _log_with_filter(fo):
+        if fo.tell() > 0:
+            if logger.logger.disabled:
+                return
+            # Filter out the error message
+            fo.seek(0)
+            suppress_output_messages = fo.readlines()
+            if record_filters:
+                filtered_messages = []
+                for i in range(len(suppress_output_messages)):
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/ortmodule/_logger.py:75]:
```suggestion
                for suppressed_message in suppress_output_messages):
```

File: orttraining/orttraining/python/training/ortmodule/_logger.py [hunk 2]
```diff
@@ -23,31 +26,66 @@ class LogLevel(IntEnum):
...
+                    found = False
+                    for warning in record_filters:
+                        if warning in suppress_output_messages[i]:
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/ortmodule/_logger.py:78]:
```suggestion
                        if warning in suppressed_message:
```

File: orttraining/orttraining/python/training/ortmodule/_logger.py [hunk 3]
```diff
@@ -23,31 +26,66 @@ class LogLevel(IntEnum):
...
+                            found = True
+                            break
+
+                    if not found:
+                        filtered_messages.append(suppress_output_messages[i])
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/ortmodule/_logger.py:83]:
```suggestion
                        filtered_messages.append(suppressed_message)
```

**Review by thiagocrepaldi — COMMENTED** [review]
Overall it looks good.

Should we add tests to make sure certain logs ARE present in certain Verbosity levels or something like that? Maybe the main most common noisy logs could be tested for selective suppression

**Comment by pengwa** [comment]
> Overall it looks good.
> 
> Should we add tests to make sure certain logs ARE present in certain Verbosity levels or something like that? Maybe the main most common noisy logs could be tested for selective suppression

Thanks for the comment! Added a unit test for the exporter warnings under different log level.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 4]
```diff
@@ -6061,3 +6061,47 @@ def generate_inputs(batch_size, max_seq_length, vocab_size):
...
+
+    N = random.randint(16, 32)  # noqa: N806
+    input = torch.randint(high=num_embeddings, size=(N,), dtype=torch.int64, device=device)
+    positions = torch.randint(high=N, size=(embedding_dim,), dtype=torch.int64, device=device)
+    _ = run_step(ort_model, input, positions)
+
+    found_missing_inference_log = False
+    for record in caplog.records:
+        msg = record.getMessage()
+        if "The shape inference of com.microsoft::SoftmaxCrossEntropyLossInternal type is missing" in msg:
```
Comment (mindest) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6100]: Looks this log is not present in VERBOSE mode either.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 5]
```diff
@@ -6061,3 +6061,52 @@ def generate_inputs(batch_size, max_seq_length, vocab_size):
     assert "ShrunkenGather" in [node.op_type for node in training_model.graph.node]
     assert "PadAndUnflatten" in [node.op_type for node in training_model.graph.node]
     del os.environ["ORTMODULE_ENABLE_EMBEDDING_SPARSE_OPTIMIZER"]
+
+
+@pytest.mark.skipif(
+    Version(torch.__version__) < Version("1.13.0"),
```
Comment (mindest) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6067]: Should be `>=`?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 4 above)
Comment (pengwa) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6105]: Thank you! For torch 1.13 or 2, there is no such warning in exporter. So restrict the tests to run only for lower version of torch.

**Review by mindest — APPROVED** [review]

File: orttraining/orttraining/python/training/ortmodule/_logger.py [hunk 6]
```diff
@@ -67,6 +110,14 @@ def ortmodule_loglevel_to_python_loglevel(loglevel: LogLevel) -> int:
     return ORTMODULE_LOG_LEVEL_MAP.get(loglevel, [Severity.WARNING, logging.WARNING])[1]
 
 
+def configure_ortmodule_logger(log_level: LogLevel) -> logging.Logger:
+    logger = logging.getLogger(f"orttraining.rank-{get_rank()}")
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_logger.py:114]: For non-distributed runs, can we avoid adding rank?

File: orttraining/orttraining/python/training/ortmodule/_logger.py [hunk 7]
```diff
@@ -23,31 +26,71 @@ class LogLevel(IntEnum):
...
+        os.dup2(old_stdout, sys.stdout.fileno())
+        os.dup2(old_stderr, sys.stderr.fileno())
+
+        if on_exit:
+            on_exit(fo)
+
+        fo.close()
+
+
+def create_log_filter(logger: logging.Logger, record_filters: Optional[List[str]], name: Optional[str]):
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_logger.py:65]: doc string for this function will help (specially for the args)

File: orttraining/orttraining/python/training/ortmodule/_logger.py [hunk 8]
```diff
@@ -23,31 +26,71 @@ class LogLevel(IntEnum):
...
 
-    suppress_logs = log_level >= LogLevel.WARNING
+    # Store original stdout and stderr file no.
+    old_stdout = os.dup(sys.stdout.fileno())
+    old_stderr = os.dup(sys.stderr.fileno())
 
-    fo = io.StringIO()
+    # Allow read and write the file.
+    fd = os.open(fo_file_path, os.O_RDWR | os.O_CREAT)
+    fo = os.fdopen(fd)
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_logger.py:44]: Can you use the a `open` context manager here?

File: orttraining/orttraining/python/training/ortmodule/_logger.py [hunk 9]
```diff
@@ -23,31 +26,71 @@ class LogLevel(IntEnum):
...
+    """Suppress output from being printed to stdout and stderr.
 
-    If there is any output detected, a single warning is issued in the context
+    If on_exit is not None, it will be called when the context manager exits.
     """
 
     # stdout and stderr is written to a tempfile instead
-    stdout = sys.stdout
-    stderr = sys.stderr
+    _, fo_file_path = tempfile.mkstemp()
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_logger.py:36]: If I am not mistaken, the temporary file created this way must be marked for deletion after work is completed. Shall we instead use the context manager?

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_logger.py:36]: good catch, man!

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 10]
```diff
@@ -177,10 +176,14 @@ def forward(self):
         pass
 
     def _build_graph(self, config):
-        if self._runtime_options.use_static_shape:
-            self._graph_builder.build(config, self._input_info.shape)
-        else:
-            self._graph_builder.build(config)
+        print_on_exit = _logger.create_log_filter(
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:179]: It is not clear that print_on_exit is a callable. Shall we call it `print_on_exit_func`?

File: orttraining/orttraining/python/training/ortmodule/_logger.py [hunk 11]
```diff
@@ -23,31 +26,71 @@ class LogLevel(IntEnum):
...
+                    if not found:
+                        filtered_messages.append(suppressed_message)
+                if filtered_messages:
+                    logger.warning(
+                        f"{'Filtered' if is_filtered else 'Full'} [{name}] logs:\n\t" + "\t".join(filtered_messages)
+                    )
+            else:
+                logger.warning(f"Full [{name}] logs:\n\t" + "\t".join(suppress_output_messages))
+
+    return _log_with_filter
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_logger.py:93]: From what I am understanding, the real reason for the outer function is so that we can bind some inputs to the inner function.
If that is the case, shall we instead use `functools.partial`?

File: orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py [hunk 12]
```diff
@@ -146,7 +146,7 @@ def initialize(self, model: ModelProto, user_input_names: List[str]) -> None:
             self._tensor_to_node_map.clear()
             for node in model.graph.node:
                 for output_name in node.output:
-                    if output_name != "":  # noqa: PLC1901
+                    if output_name != "":
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py:149]: Why is the `noqa` no longer needed? Did the linter do this automatically?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 13]
```diff
@@ -304,7 +307,17 @@ def _get_exported_model(self, input_schema: _ModelInputOutputSchemaType, *inputs
...
+
+        # VERBOSE -> FULL export verbose log + FULL torch other logs from stdout and stderr (C++ backend)
+        # INFO -> FULL export verbose log + FILTERED torch other logs from stdout and stderr (C++ backend)
+        # WARNING/ERROR -> [Rank 0] NO export verbose log + FILTERED torch other logs from stdout and stderr (C++ backend)
+        # Be noted: rank 0 log only is controlled by logger configured in _logger.py
+        torch_exporter_verbose_log = self._debug_options.logging.log_level <= LogLevel.INFO
+        print_on_exit = _logger.create_log_filter(
+            self._logger, self._debug_options.torch_exporter_filter, "model export to ONNX"
+        )
+        self._logger.warning("Exporting the PyTorch model to ONNX...")
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:319]: Is warning necessary here? It seems more like an info message.

File: orttraining/orttraining/python/training/ortmodule/_logger.py [hunk 14]
```diff
@@ -23,31 +26,71 @@ class LogLevel(IntEnum):
...
+        if fo.tell() > 0:
+            if logger.disabled:
+                return
+            # Filter out the error message
+            fo.seek(0)
+            suppress_output_messages = fo.readlines()
+            if record_filters:
+                filtered_messages = []
+                is_filtered = False
+                for suppressed_message in suppress_output_messages:
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_logger.py:76]: String matching can take a long time (specially in python).

I once found that pytorch C++ log filtering added a 2~2.5% penalty to ORTModule runs. Hoping that the filtering here will not have a negative impact on the performance.

File: orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py (hunk 12 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py:149]: yes lintrunner -a helps remove it.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 13 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: moved to info. previously I was thinking if export takes too long, we'd better let users know it is enter exporting phase.

**Review by baijumeswani — APPROVED** [review]

**Comment by pengwa** [comment]
Thanks @mindest  @baijumeswani  @thiagocrepaldi  and @justinchuby for the review!

File: orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py (hunk 12 above)
Comment (justinchuby) [reply orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py]: That said we should still prefer

if output_name:
